File: C++/AddBinary.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<unordered_set>
#include<string>
#include<string.h>
#include<algorithm>
#include<math.h>
#include<map>
using namespace std;
class Solution {
public:
	string addBinary(string a, string b) {
		int len = max(a.size(), b.size()) + 1;
		char* sum = new char[len];
		memset(sum, 0, len);
		reverse(a.begin(), a.end());
		reverse(b.begin(), b.end());
		int i = 0;
		for(i = 0; i < min(a.size(),b.size()); i++)
		{
			sum[i] += a[i] - '0' + b[i] - '0';
			if(sum[i] >= 2)
			{
				sum[i] %= 2;
				sum[i+1] += 1;
			}
		}
		while(i < a.size())
		{
			sum[i] += a[i] - '0';
			if(sum[i] >= 2)
			{
				sum[i] %= 2;
				sum[i+1] += 1;
			}
			i++;
		}
		while(i < b.size())
		{
			sum[i] += b[i] - '0';
			if(sum[i] >= 2)
			{
				sum[i] %= 2;
				sum[i+1] += 1;
			}
			i++;
		}
		if(sum[i] >= 2)
		{
			sum[i] %= 2;
			sum[i+1] += 1;
			i++;
		}
		while(sum[i] == 0)
			i--;
		if(i < 0)
			return "0";
		string ret;
		while(i >= 0)
		{
			ret.push_back(sum[i--]+'0');
		}
		delete[] sum;
		return ret;
	}
};
int main()
{
	string a = "0";
	string b = "1";
	Solution sol;
	cout<<sol.addBinary(a, b)<<endl;
	getchar();
}

File: JAVA/AddBinary.java
class Solution {
    public String addBinary(String a, String b) {
        // Start typing your Java solution below
        // DO NOT write main() function
    	if(a == null || b == null)
    		return null;
        a = reverse(a);
        b = reverse(b);
        StringBuilder sb = new StringBuilder();
        int len = Math.min(a.length(), b.length());
        int sum = 0;
        int carry = 0;
        int i = 0;
        for(; i < len; i++)
        {
        	sum = a.charAt(i)-'0' + b.charAt(i)-'0' + carry;
        	carry = sum/2;
        	sum %= 2;
        	sb.append(String.valueOf(sum));
        }
        while(i < a.length())
        {
        	sum = a.charAt(i)-'0' + carry;
        	carry = sum/2;
        	sum %= 2;
        	sb.append(String.valueOf(sum));
        	i++;
        }
        while(i < b.length())
        {
        	sum = b.charAt(i)-'0' + carry;
        	carry = sum/2;
        	sum %= 2;
        	sb.append(String.valueOf(sum));
        	i++;
        }
        if(carry == 1)
        	sb.append('1');
        return reverse(sb.toString());
    }

	private String reverse(String s) {
		StringBuilder sb = new StringBuilder();
		int len = s.length();
		for(int i = len - 1; i >= 0; i--)
		{
			sb.append(s.charAt(i));
		}
		return sb.toString();
	}
}
public class AddBinary {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
